File: frontend/api/feed.ts
export type FeedItem = {
  id: string
  hlsUrl: string
  thumbnailUrl?: string
  caption?: string
  creator?: { id: string; name: string }
  createdAt: string
}

export type FeedResponse = {
  items: FeedItem[]
  nextCursor: string | null
}

const API_BASE = process.env.EXPO_PUBLIC_API_BASE ?? 'http://localhost:3000'

export async function fetchFeed(params: {
  cursor?: string | null
  limit?: number
}): Promise<FeedResponse> {
  const { cursor, limit = 10 } = params
  const url = new URL(`${API_BASE}/feed`)

  url.searchParams.set('limit', String(limit))

  if (cursor) {
    url.searchParams.set('cursor', cursor)
  }

  const result = await fetch(url.toString(), {
    headers: {
      Accept: 'application/json',
    },
  })

  if (!result.ok) {
    const text = await result.text().catch(() => '')
    throw new Error(`feed fetch failed: ${result.status} ${text}`)
  }

  return (await result.json()) as FeedResponse
}

File: backend/crates/shared/src/dto/feed.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Deserialize)]
pub struct FeedQuery {
  pub cursor: Option<String>,
  pub limit: Option<i64>,
}

#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct FeedResponse {
  pub items: Vec<FeedItem>,
  pub next_cursor: Option<String>,
}

#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct FeedItem {
  pub id: String,
  pub hls_url: String,
  pub thumbnail_url: Option<String>,
  pub caption: Option<String>,
  pub creator: Creator,
  pub created_at: String,
}

#[derive(Debug, Serialize)]
pub struct Creator {
  pub id: String,
  pub name: String,
}
